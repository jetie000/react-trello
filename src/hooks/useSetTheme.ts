

export const useSetTheme = () => {
  const setTheme = () => {
    document.documentElement.setAttribute(
      "data-bs-theme",
      localStorage.getItem(localStorage.THEME_LOCALSTORAGE) || "dark"
    )
  }
  return setTheme
}
