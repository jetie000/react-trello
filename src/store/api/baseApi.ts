import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { variables } from "@/variables";
import { api } from "@/http";

const axiosBaseQuery =
  ({ baseUrl } = { baseUrl: '' }) =>
    async ({ url, method, data, params, headers }: any) => {
      try {
        const result = await api({
          url: baseUrl + url,
          method,
          data,
          params,
          headers,
        })
        return { data: result.data }
      } catch (axiosError: any) {
        const err = axiosError
        return {
          error: {
            status: err.response?.status,
            data: err.response?.data || err.message,
          },
        }
      }
    }

export const baseApi = createApi({
  reducerPath: 'baseApi',
  baseQuery: axiosBaseQuery({
    baseUrl: variables.API_URL,
  }),
  tagTypes: ['User', 'Users'],
  endpoints: () => ({}),
});