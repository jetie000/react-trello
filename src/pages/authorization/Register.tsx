import { useEffect, useRef } from "react"
import { useRegisterUserMutation } from "@/store/api/user.api"
import { useActions } from "@/hooks/useActions"
import { IError } from "@/types/error.interface"
import { FetchBaseQueryError } from "@reduxjs/toolkit/query"
import { RootStateStore } from "@/store/store"
import { useSelector } from "react-redux"
import { languages } from "@/config/languages"

function Register() {
  const [registerUser, { isLoading, isSuccess, isError, error }] = useRegisterUserMutation()
  const { language } = useSelector((state: RootStateStore) => state.options)
  const { showToast } = useActions()
  const fullNameRef = useRef<HTMLInputElement>(null)
  const emailRef = useRef<HTMLInputElement>(null)
  const passwordRef = useRef<HTMLInputElement>(null)

  useEffect(() => {
    if (isSuccess && fullNameRef.current && passwordRef.current && emailRef.current) {
      showToast(languages[language].SUCCESFULLY_REGISTERED)
      fullNameRef.current.value = ""
      passwordRef.current.value = ""
      emailRef.current.value = ""
    }
    if (isError) {
      showToast(((error as FetchBaseQueryError)?.data as IError).message)
    }
  }, [isLoading])

  const isValidFields = () =>
    fullNameRef.current &&
    passwordRef.current &&
    emailRef.current &&
    fullNameRef.current.value !== "" &&
    passwordRef.current.value !== "" &&
    emailRef.current.value !== ""

  const registerClick = () => {
    if (isLoading) return
    if (isValidFields()) {
      registerUser({
        email: emailRef.current!.value.trim(),
        password: passwordRef.current!.value.trim(),
        fullName: fullNameRef.current!.value.trim()
      })
      return
    }
    showToast(languages[language].INPUT_DATA)
  }

  return (
    <div className="mt-3">
      <form>
        <div className="mb-3">
          <label htmlFor="inputFullName">{languages[language].SURNAME_AND_NAME}</label>
          <input
            className="form-control"
            id="inputFullName"
            placeholder={languages[language].ENTER_SURNAME_NAME}
            ref={fullNameRef}
          />
        </div>
        <div className="mb-3">
          <label htmlFor="inputEmail">Email</label>
          <input
            type="email"
            className="form-control"
            id="inputEmail"
            placeholder={languages[language].ENTER_EMAIL}
            ref={emailRef}
          />
        </div>
        <div className="mb-3">
          <label htmlFor="inputPassword">{languages[language].PASSWORD}</label>
          <input
            type="password"
            className="form-control"
            id="inputPassword"
            placeholder={languages[language].ENTER_PASSWORD}
            ref={passwordRef}
          />
        </div>
        <button type="button" className="btn btn-primary mt-3 w-100" onClick={registerClick}>
          {isLoading ? (
            <div className="spinner-border spinner-border-sm" role="status">
              <span className="visually-hidden">{languages[language].LOADING}</span>
            </div>
          ) : (
            languages[language].REGISTER_
          )}
        </button>
      </form>
    </div>
  )
}

export default Register
