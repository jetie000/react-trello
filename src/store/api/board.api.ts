import { baseApi } from "./baseApi"
import { IError } from "@/types/error.interface"
import { IBoard, IBoardAddInfo, IBoardUpdateInfo } from "@/types/board.interface"

export const boardApi = baseApi.injectEndpoints({
  endpoints: builder => ({
    getBoardById: builder.query<IBoard | IError, number>({
      query: (id: number) => ({
        url: "/board/" + id,
        method: "GET"
      }),
      providesTags: ["Board"]
    }),
    getBoardByUserId: builder.query<IBoard[] | IError, number>({
      query: (id: number) => ({
        url: "/board/user/" + id,
        method: "GET"
      }),
      providesTags: ["Boards"]
    }),
    addBoard: builder.mutation<IBoard | IError, IBoardAddInfo>({
      query: (boardInfo: IBoardAddInfo) => ({
        body: boardInfo,
        url: "/board",
        method: "POST"
      }),
      invalidatesTags: ["Boards"]
    }),
    changeBoard: builder.mutation<IBoard | IError, IBoardUpdateInfo>({
      query: (boardInfo: IBoardUpdateInfo) => ({
        body: boardInfo,
        url: "/board",
        method: "PUT"
      }),
      invalidatesTags: ["Board", "Boards"]
    }),
    deleteBoard: builder.mutation<IBoard | IError, number>({
      query: (boardId: number) => ({
        url: "/board/" + boardId,
        method: "DELETE"
      }),
      invalidatesTags: ["Board", "Boards"]
    }),
    leaveBoard: builder.mutation<IBoard | IError, number>({
      query: (boardId: number) => ({
        url: "/board/" + boardId + "/leave",
        method: "POST"
      }),
      invalidatesTags: ["Board", "Boards"]
    })
  })
})

export const {
  useAddBoardMutation,
  useChangeBoardMutation,
  useDeleteBoardMutation,
  useGetBoardByIdQuery,
  useGetBoardByUserIdQuery,
  useLeaveBoardMutation
} = boardApi
