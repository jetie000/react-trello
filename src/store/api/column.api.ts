import { baseApi } from './baseApi'
import { IError } from '@/types/error.interface'
import { IColumn, IColumnAddInfo, IColumnUpdateInfo } from '@/types/column.interface'

export const columnApi = baseApi.injectEndpoints({
    endpoints: builder => ({
        addColumn: builder.mutation<IColumn | IError, IColumnAddInfo>({
            query: (columnInfo: IColumnAddInfo) => ({
                body: columnInfo,
                url: '/column',
                method: 'POST',
            }),
        }),
        changeColumn: builder.mutation<IColumn | IError, IColumnUpdateInfo>({
            query: (columnInfo: IColumnUpdateInfo) => ({
                body: columnInfo,
                url: '/column',
                method: 'PUT',
            }),
        }),
        deleteColumn: builder.mutation<IColumn | IError, number>({
            query: (columnId: number) => ({
                url: '/column/'+columnId,
                method: 'DELETE',
            })
        }),
    })
})

export const {
    useAddColumnMutation,
    useChangeColumnMutation,
    useDeleteColumnMutation
 } = columnApi