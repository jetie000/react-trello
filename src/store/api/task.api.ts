import { baseApi } from './baseApi'
import { IError } from '@/types/error.interface'
import { ITask, ITaskAddInfo, ITaskUpdateInfo } from '@/types/task.interface'

export const taskApi = baseApi.injectEndpoints({
    endpoints: builder => ({
        addTask: builder.mutation<ITask | IError, ITaskAddInfo>({
            query: (taskInfo: ITaskAddInfo) => ({
                body: taskInfo,
                url: '/task',
                method: 'POST',
            }),
        }),
        changeTask: builder.mutation<ITask | IError, ITaskUpdateInfo>({
            query: (taskInfo: ITaskUpdateInfo) => ({
                body: taskInfo,
                url: '/task',
                method: 'PUT',
            }),
        }),
        deleteTask: builder.mutation<ITask | IError, number>({
            query: (taskId: number) => ({
                url: '/task/'+taskId,
                method: 'DELETE',
            })
        }),
    })
})

export const {
 } = taskApi