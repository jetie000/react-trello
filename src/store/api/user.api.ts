import { AuthResponse } from '@/types/authResponse.interface'
import { baseApi } from './baseApi'
import { IUserChangeInfo, IUserLoginInfo, IUserRegisterInfo, IUserResponse} from '@/types/user.interface'

export const userApi = baseApi.injectEndpoints({
    endpoints: builder => ({
        getByIds: builder.query<IUserResponse[], string>({
            query: (ids: string) => ({
                url: '/user/getByIds?ids='+ids,
                method: 'POST',
            })
        }),
        getById: builder.query<AuthResponse | string, number>({
            query: (id: number) => ({
                url: '/user/getById?id='+id,
                method: 'POST',
            })
        }),
        logInUser: builder.mutation<AuthResponse | string, IUserLoginInfo>({
            query: (userInfo: IUserLoginInfo) => ({
                body: userInfo,
                url: '/auth/login',
                method: 'POST',
            })
        }),
        registerUser: builder.mutation<AuthResponse | string, IUserRegisterInfo>({
            query: (userInfo: IUserRegisterInfo) => ({
                body: userInfo,
                url: '/auth/register',
                method: 'POST',
            }),
        }),
        changeUser: builder.mutation<AuthResponse | string, IUserChangeInfo>({
            query: (userInfo: IUserChangeInfo) => ({
                body: userInfo,
                url: '/user',
                method: 'PUT',
            }),
        }),
        logout: builder.mutation<string, null>({
            query: () => ({
                url: '/auth/logout',
                method: 'POST',
            }),
        }),
        deleteUser: builder.mutation<string, number>({
            query: (userId: number) => ({
                url: '/user/'+userId,
                method: 'DELETE',
            })
        }),
    })
})

export const { } = userApi