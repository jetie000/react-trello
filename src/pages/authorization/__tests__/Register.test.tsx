import React from "react"
import * as reduxHooks from "react-redux"
import * as reactRouterDom from "react-router-dom"
import * as actions from "@/hooks/useActions"
import * as userApi from "@/store/api/user.api"
import { fireEvent, render, screen } from "@testing-library/react"
import { AuthResponse } from "@/types/authResponse.interface"
import Register from "../Register"

jest.mock("@/store/store")
jest.mock("react-redux")
jest.mock("react-router-dom")
jest.mock("@/hooks/useActions")
jest.mock("@/store/api/user.api")
const mockUseSelector = jest.spyOn(reduxHooks, "useSelector")
const mockUseActions = jest.spyOn(actions, "useActions")
const mockRegister = jest.spyOn(userApi, "useRegisterUserMutation")
const mockUseNavigate = jest.spyOn(reactRouterDom, "useNavigate")

const mockReturnSelectorValue = {
  language: 1,
  id: 1,
  token: "token"
}
const mockNavigateReturn = jest.fn()
const mockRegisterFunc = jest.fn()
const showToastMock = jest.fn()
const mockReturnActionsValue = { showToast: showToastMock, login: jest.fn() }

const mockReturnRegister = (isLoading: boolean, isPressed: boolean) => ({
  isLoading: isLoading && isPressed,
  isError: false,
  isSuccess: !isLoading && isPressed,
  refetch: jest.fn(),
  data:
    !isLoading && isPressed
      ? ({
          accessToken: "11",
          refreshToken: "22",
          email: "test@example.com"
        } as AuthResponse)
      : undefined,
  reset: jest.fn()
})

describe("Register", () => {
  beforeEach(() => {
    jest.restoreAllMocks()
    jest.resetAllMocks()
    mockUseSelector.mockReturnValue(mockReturnSelectorValue)
    mockUseActions.mockReturnValue(mockReturnActionsValue as any)
    mockRegister.mockReturnValue([mockRegisterFunc, mockReturnRegister(false, false)])
    mockUseNavigate.mockReturnValue(mockNavigateReturn)
  })
  it("should register user when all fields are filled correctly", () => {
    const component = render(<Register />)
    expect(mockUseActions).toHaveBeenCalled()
    expect(mockRegister).toHaveBeenCalled()
    expect(mockUseSelector).toHaveBeenCalled()

    fireEvent.change(screen.getByPlaceholderText("Enter surname and name"), {
      target: { value: "John Doe" }
    })
    fireEvent.change(screen.getByPlaceholderText("Input e-mail"), {
      target: { value: "john.doe@example.com" }
    })
    fireEvent.change(screen.getByPlaceholderText("Input password"), {
      target: { value: "password123" }
    })
    fireEvent.click(screen.getByRole("button", { name: "Sign Up" }))
    mockRegister.mockReturnValue([mockRegisterFunc, mockReturnRegister(true, true)])
    component.rerender(<Register />)
    mockRegister.mockReturnValue([mockRegisterFunc, mockReturnRegister(false, true)])
    component.rerender(<Register />)
    expect(mockRegisterFunc).toHaveBeenCalledWith({
      email: "john.doe@example.com",
      password: "password123",
      fullName: "John Doe"
    })
    expect(showToastMock).toHaveBeenCalledWith(
      "You've succesfully registered. Check your e-mail for confirm letter"
    )
    expect(showToastMock).toHaveBeenCalledTimes(1)
    expect(component).toMatchSnapshot()
  })
  it("should show enter data message when provided invalid data", () => {
    render(<Register />)

    fireEvent.click(screen.getByRole("button", { name: "Sign Up" }))
    expect(showToastMock).toHaveBeenCalledWith("Input data")
    expect(showToastMock).toHaveBeenCalledTimes(1)
  })
  it("should show loading spinner while fetching", () => {
    const component = render(<Register />)

    fireEvent.change(screen.getByPlaceholderText("Enter surname and name"), {
      target: { value: "John Doe" }
    })
    fireEvent.change(screen.getByPlaceholderText("Input e-mail"), {
      target: { value: "john.doe@example.com" }
    })
    fireEvent.change(screen.getByPlaceholderText("Input password"), {
      target: { value: "password123" }
    })
    fireEvent.click(screen.getByRole("button", { name: "Sign Up" }))
    mockRegister.mockReturnValue([mockRegisterFunc, mockReturnRegister(true, true)])
    component.rerender(<Register />)
    expect(screen.getByText(/Loading/i)).toBeInTheDocument()
    expect(component).toMatchSnapshot()
  })
})
