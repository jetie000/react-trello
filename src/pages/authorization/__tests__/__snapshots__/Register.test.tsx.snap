// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register should register user when all fields are filled correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mt-3"
      >
        <form>
          <div
            class="mb-3"
          >
            <label
              for="inputFullName"
            >
              Surname and name
            </label>
            <input
              class="form-control"
              id="inputFullName"
              placeholder="Enter surname and name"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              for="inputEmail"
            >
              Email
            </label>
            <input
              class="form-control"
              id="inputEmail"
              placeholder="Input e-mail"
              type="email"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              for="inputPassword"
            >
              Password
            </label>
            <input
              class="form-control"
              id="inputPassword"
              placeholder="Input password"
              type="password"
            />
          </div>
          <button
            class="btn btn-primary mt-3 w-100"
            type="button"
          >
            Sign Up
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mt-3"
    >
      <form>
        <div
          class="mb-3"
        >
          <label
            for="inputFullName"
          >
            Surname and name
          </label>
          <input
            class="form-control"
            id="inputFullName"
            placeholder="Enter surname and name"
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            for="inputEmail"
          >
            Email
          </label>
          <input
            class="form-control"
            id="inputEmail"
            placeholder="Input e-mail"
            type="email"
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            for="inputPassword"
          >
            Password
          </label>
          <input
            class="form-control"
            id="inputPassword"
            placeholder="Input password"
            type="password"
          />
        </div>
        <button
          class="btn btn-primary mt-3 w-100"
          type="button"
        >
          Sign Up
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Register should show loading spinner while fetching 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mt-3"
      >
        <form>
          <div
            class="mb-3"
          >
            <label
              for="inputFullName"
            >
              Surname and name
            </label>
            <input
              class="form-control"
              id="inputFullName"
              placeholder="Enter surname and name"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              for="inputEmail"
            >
              Email
            </label>
            <input
              class="form-control"
              id="inputEmail"
              placeholder="Input e-mail"
              type="email"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              for="inputPassword"
            >
              Password
            </label>
            <input
              class="form-control"
              id="inputPassword"
              placeholder="Input password"
              type="password"
            />
          </div>
          <button
            class="btn btn-primary mt-3 w-100"
            type="button"
          >
            <div
              class="spinner-border spinner-border-sm"
              role="status"
            >
              <span
                class="visually-hidden"
              >
                Loading...
              </span>
            </div>
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mt-3"
    >
      <form>
        <div
          class="mb-3"
        >
          <label
            for="inputFullName"
          >
            Surname and name
          </label>
          <input
            class="form-control"
            id="inputFullName"
            placeholder="Enter surname and name"
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            for="inputEmail"
          >
            Email
          </label>
          <input
            class="form-control"
            id="inputEmail"
            placeholder="Input e-mail"
            type="email"
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            for="inputPassword"
          >
            Password
          </label>
          <input
            class="form-control"
            id="inputPassword"
            placeholder="Input password"
            type="password"
          />
        </div>
        <button
          class="btn btn-primary mt-3 w-100"
          type="button"
        >
          <div
            class="spinner-border spinner-border-sm"
            role="status"
          >
            <span
              class="visually-hidden"
            >
              Loading...
            </span>
          </div>
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
